src/humanloop/decorators/prompt.py

```diff
@@ -0,0 +1,149 @@
+                raise ValueError(f"{func.__name__}: Frequency penalty parameter must be between -2 and 2")
+            prompt_kernel["frequency_penalty"] = frequency_penalty
+
+        for attr_name, attr_value in {
+            "model": model,
+            "endpoint": endpoint,
+            "template": template,
+            "provider": provider,
+            "max_tokens": max_tokens,
+            "stop": stop,
+            "other": other,
+            "seed": seed,
+            "response_format": response_format,
+            # {} -> None
+            "attributes": attributes or None,
+            "tools": tools,
+        }.items():
+            prompt_kernel[attr_name] = attr_value  # type: ignore
+
+        @wraps(func)
+        def wrapper(*args: Sequence[Any], **kwargs: Mapping[str, Any]) -> Any:
+            span: Span
+            with opentelemetry_tracer.start_as_current_span(str(uuid.uuid4())) as span:
+                span_id = span.get_span_context().span_id
+                if span.parent:
+                    span_parent_id = span.parent.span_id
+                    parent_trace_metadata = TRACE_FLOW_CONTEXT.get(span_parent_id, {})
+                    if parent_trace_metadata:
+                        TRACE_FLOW_CONTEXT[span_id] = FlowContext(
+                            trace_id=parent_trace_metadata["trace_id"],
+                            trace_parent_id=span_parent_id,
+                            is_flow_log=False,
+                        )
+
+                span.set_attribute(HL_PATH_KEY, path if path else func.__name__)
+                span.set_attribute(HL_FILE_TYPE_KEY, "prompt")
+
+                if prompt_kernel:
+                    write_to_opentelemetry_span(
+                        span=span,
+                        key=f"{HL_FILE_KEY}.prompt",
+                        value=prompt_kernel,  # type: ignore
+                    )
+
+                # Call the decorated function
+                try:
+                    output = func(*args, **kwargs)
+                    error = None
+                except Exception as e:
+                    logger.error(f"Error calling {func.__name__}: {e}")
+                    output = None
+                    error = str(e)
+
+                prompt_log = {
+                    "inputs": args_to_inputs(func, args, kwargs),
+                    "output": output,
+                    "error": error,
+                }
+                write_to_opentelemetry_span(
+                    span=span,
+                    key=HL_LOG_KEY,
+                    value=prompt_log,
+                )
+
+            # Return the output of the decorated function
+            return output
+
+        prompt_kernel_file = {**prompt_kernel}
+        if prompt_kernel_file.get("provider") is None:
+            prompt_kernel_file["provider"] = "openai"  # type: ignore
+        if prompt_kernel_file.get("endpoint") is None:
+            prompt_kernel_file["endpoint"] = "chat"  # type: ignore
```

Comment:
why are we manually applying defaults here like this? surely there's a better way if this needs to be done?

Reply:
This is tied to another issue, I will address in future iteration
https://linear.app/humanloop/issue/ENG-1350/fix-prompt-upsert-discrepancy

Reply:
Add a comment. (and please be precise about what the issue is)